### PR TITLE
Wait for rc store synced before syncing deployment

### PR DESCRIPTION
**Pull request by janetkuo**
Attempt to fix flaky deployment e2e tests #18996 #19064 #19299

In the failed cases, multiple rcs are created by a deployment and messes things up. Now we make sure rc store is synced before syncing deployment. 

cc @nikhiljindal @ghodss @ironcladlou @kargakis @bgrant0607

### CODE REVIEW

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -375,6 +388,14 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return nil
 	}
 	d := *obj.(*extensions.Deployment)
+	if !dc.rcStoreSynced() {
```
Comment (nikhiljindal) [inline pkg/controller/deployment/deployment_controller.go]: Do we need to wait for the whole store to sync!
What if just check that old and new RCs dont have any unsatisfied expectation?

**Comment by nikhiljindal** [comment]
As discussed, this should fix the test, but we can optimize it by using expectations.

I am fine submitting this PR as is to fix 3 P0s!
In another PR, we can add expectations for RCs and use that.

**Comment by ironcladlou** [comment]
Agree that we should move forward with a consistent but inefficient solution now with a followup to optimize.

**Comment by janetkuo** [comment]
@k8s-oncall Can we manually merge this to see if it solves test flake?

**Pull request merged**